src/client/datascience/variablesView/variableViewNotebookWatcher.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+import { inject, injectable } from 'inversify';
+import { Event, EventEmitter } from 'vscode';
+import '../../common/extensions';
+import { IFileSystem } from '../../common/platform/types';
+import { IDisposableRegistry } from '../../common/types';
+import { KernelState, KernelStateEventArgs } from '../notebookExtensibility';
+import { INotebook, INotebookEditor, INotebookEditorProvider, INotebookExtensibility } from '../types';
+import { IVariableViewNotebookWatcher } from './types';
+
+// For any class that is monitoring the variables of the active notebook document, this class will update you
+// when the active notebook changes or if the execution count is updated on the active notebook
+// NOTE: Currently this class is only looking at native notebook documents
+@injectable()
+export class VariableViewNotebookWatcher implements IVariableViewNotebookWatcher {
+    public get onDidChangeActiveVariableViewNotebook(): Event<INotebook | undefined> {
+        return this._onDidChangeActiveVariableViewNotebook.event;
+    }
+    public get onDidExecuteActiveVariableViewNotebook(): Event<{ executionCount: number }> {
+        return this._onDidExecuteActiveVariableViewNotebook.event;
+    }
+    public get activeVariableViewNotebook(): INotebook | undefined {
+        return this.notebookEditorProvider.activeEditor?.notebook;
+    }
+
+    private readonly _onDidExecuteActiveVariableViewNotebook = new EventEmitter<{ executionCount: number }>();
+    private readonly _onDidChangeActiveVariableViewNotebook = new EventEmitter<INotebook | undefined>();
+
+    constructor(
+        @inject(INotebookEditorProvider) private readonly notebookEditorProvider: INotebookEditorProvider,
+        @inject(INotebookExtensibility) private readonly notebookExtensibility: INotebookExtensibility,
+        @inject(IFileSystem) private readonly fileSystem: IFileSystem,
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry
+    ) {
+        // We need to know if kernel state changes or if the active notebook editor is changed
+        this.notebookExtensibility.onKernelStateChange(this.kernelStateChanged, this, this.disposables);
+        this.notebookEditorProvider.onDidChangeActiveNotebookEditor(this.activeEditorChanged, this, this.disposables);
+    }
+
+    // When the kernel state is changed we need to see if it's a cell from the active document that finished execution
+    // If so update the execution count on the variable view to refresh variables
+    private async kernelStateChanged(kernelStateEvent: KernelStateEventArgs) {
```

Comment:
Code is the same as what used to be in variableview, but now it just triggers the EventEmitter instead of posting a message to the UI.